### PR TITLE
Pricing 4th phase

### PR DESCRIPTION
**Pull request by ClaudeROSSI**
**Pricing Engine - First Implementation.**

Activated with the feature toggle: Feature.PRICING_NEW_MODEL
New invoice item presentation: Feature.BILLING_SHOW_SESSION_HEADER

You can switch them OFF to go back to the previous behavior ==> Simple Pricing billing only the Energy

### CODE REVIEW

File: src/integration/billing/stripe/StripeBillingIntegration.ts [hunk 1]
```diff
@@ -805,79 +821,83 @@ export default class StripeBillingIntegration extends BillingIntegration {
     };
   }
 
-  private _buildInvoiceItemParkTimeParameters(customerID: string, billingInvoiceItem: BillingInvoiceItem, invoiceID?: string): Stripe.InvoiceItemCreateParams {
-    const { parkingData, taxes } = billingInvoiceItem;
-    const currency = parkingData.pricingData.currency.toLowerCase();
+  private async _createStripeInvoiceItems(customerID: string, billingInvoiceItem: BillingInvoiceItem, invoiceID?: string): Promise<void> {
```
Comment (LucasBrazi06) [inline src/integration/billing/stripe/StripeBillingIntegration.ts:824]: US in private method ;-)

File: src/integration/billing/stripe/StripeBillingIntegration.ts [hunk 2]
```diff
@@ -899,7 +919,8 @@ export default class StripeBillingIntegration extends BillingIntegration {
           user: transaction.userID,
           action: ServerAction.BILLING_TRANSACTION,
           module: MODULE_NAME, method: 'stopTransaction',
-          message: `Transaction data is suspicious - billing operation has been aborted - transaction ID: ${transaction.id}`
+          message: `Transaction data is suspicious - billing operation has been aborted - transaction ID: ${transaction.id}`,
+          ...LoggingHelper.getSessionProperties(transaction)
```
Comment (LucasBrazi06) [inline src/integration/billing/stripe/StripeBillingIntegration.ts:923]: getSessionProperties -> getTransactionProperties?

File: src/integration/billing/stripe/StripeBillingIntegration.ts [hunk 3]
```diff
@@ -995,69 +1018,141 @@ export default class StripeBillingIntegration extends BillingIntegration {
...
+    }
+    const parkingTime = pricingConsumptionData.parkingTime;
+    if (parkingTime) {
+      parkingTime.itemDescription = this.buildLineItemDescription4PricedDimension(transaction, DimensionType.PARKING_TIME, parkingTime);
+      parkingTime.taxes = taxes;
+    }
+    return pricingConsumptionData;
+  }
+
+  private _convertToBillingInvoiceItem(transaction: Transaction) : BillingInvoiceItem {
```
Comment (LucasBrazi06) [inline src/integration/billing/stripe/StripeBillingIntegration.ts:1115]: US

Comment (LucasBrazi06) [inline src/integration/billing/stripe/StripeBillingIntegration.ts:1115]: private after public methods, eslint should have yelled :-)

File: src/integration/billing/stripe/StripeHelpers.ts [hunk 4]
```diff
@@ -1,4 +1,4 @@
-import { BillingAdditionalData, BillingError, BillingErrorCode, BillingErrorType, BillingInvoice, BillingInvoiceItem, BillingOperationResult, BillingSessionData } from '../../../types/Billing';
+import { BillingAdditionalData, BillingError, BillingErrorCode, BillingErrorType, BillingInvoice, BillingInvoiceItem, BillingSessionData } from '../../../types/Billing';
```
Comment (LucasBrazi06) [inline src/integration/billing/stripe/StripeHelpers.ts:1]: Class name in singular?

File: src/integration/pricing/ConsumptionPricer.ts [hunk 5]
```diff
@@ -0,0 +1,364 @@
+import { DimensionType, PricedConsumptionData, PricedDimensionData, PricingDimension, PricingRestriction, ResolvedPricingDefinition, ResolvedPricingModel } from '../../types/Pricing';
+
+import Consumption from '../../types/Consumption';
+import Tenant from '../../types/Tenant';
+import Utils from '../../utils/Utils';
+import moment from 'moment';
+
+export default class ConsumptionPricer {
+
+  tenant: Tenant;
```
Comment (LucasBrazi06) [inline src/integration/pricing/ConsumptionPricer.ts:10]: No class modifiers? private?

File: src/integration/pricing/ConsumptionPricer.ts [hunk 6]
```diff
@@ -0,0 +1,364 @@
+import { DimensionType, PricedConsumptionData, PricedDimensionData, PricingDimension, PricingRestriction, ResolvedPricingDefinition, ResolvedPricingModel } from '../../types/Pricing';
+
+import Consumption from '../../types/Consumption';
+import Tenant from '../../types/Tenant';
+import Utils from '../../utils/Utils';
+import moment from 'moment';
+
+export default class ConsumptionPricer {
+
```
Comment (LucasBrazi06) [inline src/integration/pricing/ConsumptionPricer.ts:9]: BL

File: src/integration/pricing/PricingEngine.ts [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+    await Logging.logInfo({
+      tenantID: tenant.id,
+      module: MODULE_NAME,
+      action: ServerAction.PRICING,
+      method: 'resolvePricingContext',
+      message: `Pricing context has been resolved - ${pricingDefinitions.length} pricing definitions have been found`,
+      detailedMessages: { resolvedPricingModel },
+      ...LoggingHelper.getSessionProperties(transaction)
+    });
+    return Promise.resolve(resolvedPricingModel);
```
Comment (LucasBrazi06) [inline src/integration/pricing/PricingEngine.ts:52]: A simple return of resolvedPricingModel should suffice

File: src/integration/pricing/PricingEngine.ts [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+      module: MODULE_NAME,
+      action: ServerAction.PRICING,
+      method: 'getPricingDefinitions4Entity',
+      message: `Pricing context resolution - ${actualPricingDefinitions.length || 0} pricing definitions found for ${entityType}: '${entityID}'`,
+      ...LoggingHelper.getSessionProperties(transaction)
+    });
+    return actualPricingDefinitions || [];
+  }
+
+  private static async _getPricingDefinitions4Entity(tenant: Tenant, entityID: string): Promise<PricingDefinition[]> {
```
Comment (LucasBrazi06) [inline src/integration/pricing/PricingEngine.ts:90]: US

File: src/integration/pricing/PricingEngine.ts [hunk 9]
```diff
@@ -0,0 +1,185 @@
...
+
+  private static checkEntityType(pricingDefinition: PricingDefinition, entityType: PricingEntity) : PricingDefinition {
+    return (pricingDefinition.entityType === entityType) ? pricingDefinition : null;
+  }
+
+  private static checkStaticRestrictions(pricingDefinition: PricingDefinition, transaction: Transaction, chargingStation: ChargingStation) : PricingDefinition {
+    if (pricingDefinition.staticRestrictions) {
+      if (
+        !PricingEngine.checkDateValidity(pricingDefinition.staticRestrictions, transaction)
+      || !PricingEngine.checkConnectorType(pricingDefinition.staticRestrictions, transaction, chargingStation)
```
Comment (LucasBrazi06) [inline src/integration/pricing/PricingEngine.ts:111]: Operators || at the end of the previous line

File: src/integration/pricing/PricingFactory.ts [hunk 10]
```diff
@@ -1,6 +1,8 @@
+import FeatureToggles, { Feature } from '../../utils/FeatureToggles';
 import { PricingSetting, PricingSettingsType } from '../../types/Setting';
 import Tenant, { TenantComponents } from '../../types/Tenant';
```
Comment (LucasBrazi06) [inline src/integration/pricing/PricingFactory.ts:4]: BL

File: src/integration/pricing/PricingHelper.ts [hunk 11]
```diff
@@ -0,0 +1,65 @@
+/* eslint-disable @typescript-eslint/member-ordering */
+/* eslint-disable max-len */
+import { DimensionType, PricedConsumptionData, PricedDimensionData, PricingDimensions } from '../../types/Pricing';
+
+import Decimal from 'decimal.js';
+import Utils from '../../utils/Utils';
+
+export default class PricingHelper {
+
+  public static accumulatePricingDimensions(pricingDimensions: PricingDimensions[]): { cumulatedAmount: number, cumulatedRoundedAmount: number } {
```
Comment (LucasBrazi06) [inline src/integration/pricing/PricingHelper.ts:10]: Create a type for this?
{ cumulatedAmount: number, cumulatedRoundedAmount: number }

File: src/server/rest/v1/service/PricingService.ts [hunk 12]
```diff
@@ -0,0 +1,175 @@
...
+
+export default class PricingService {
+
+  public static async handleGetPricingDefinition(action: ServerAction, req: Request, res: Response, next: NextFunction): Promise<void> {
+    // Check if component is active
+    UtilsService.assertComponentIsActiveFromToken(req.user, TenantComponents.PRICING,
+      Action.READ, Entity.PRICING_DEFINITION, MODULE_NAME, 'handleGetPricingDefinition');
+    // Filter
+    const filteredRequest = PricingValidator.getInstance().validatePricingDefinitionGet(req.query);
+    UtilsService.assertIdIsProvided(action, filteredRequest.ID, MODULE_NAME, 'handleGetPricingDefinition', req.user);
```
Comment (LucasBrazi06) [inline src/server/rest/v1/service/PricingService.ts:27]: To remove, this is checked already in UtilsService.checkAndGetPricingDefinitionAuthorization()

File: src/storage/mongodb/PricingStorage.ts [hunk 13]
```diff
@@ -0,0 +1,208 @@
...
+    } else {
+      entityID = DatabaseUtils.convertToObjectID(pricingDefinition.entityID);
+    }
+    const pricingDefinitionMDB = {
+      _id: pricingDefinition.id ? DatabaseUtils.convertToObjectID(pricingDefinition.id) : new ObjectId(),
+      entityID,
+      entityType: pricingDefinition.entityType,
+      name: pricingDefinition.name,
+      description: pricingDefinition.description,
+      staticRestrictions: pricingDefinition.staticRestrictions,
```
Comment (LucasBrazi06) [inline src/storage/mongodb/PricingStorage.ts:34]: Refine each type up to the bone?

File: src/storage/mongodb/PricingStorage.ts [hunk 14]
```diff
@@ -0,0 +1,208 @@
...
+      pricingDefinitionIDs: [id],
+      entityIDs: params.entityIDs,
+      entityTypes: params.entityTypes,
+      withEntityInformation: params.withEntityInformation
+    }, Constants.DB_PARAMS_SINGLE_RECORD, projectFields);
+    return pricingDefinitionMDB.count === 1 ? pricingDefinitionMDB.result[0] : null;
+  }
+
+  public static async getPricingDefinitions(tenant: Tenant,
+      params: {
```
Comment (LucasBrazi06) [inline src/storage/mongodb/PricingStorage.ts:77]: OL

File: src/storage/mongodb/PricingStorage.ts [hunk 15]
```diff
@@ -137,10 +145,31 @@ export default class PricingStorage {
     aggregation.push({
       $limit: dbParams.limit
     });
+    // Sites
+    if (params.withEntityInformation) {
```
[the remark this answers is not in the document]
Comment (LucasBrazi06) [reply src/storage/mongodb/PricingStorage.ts:149]: Ok

File: src/types/Authorization.ts [hunk 16]
```diff
@@ -84,6 +84,8 @@ export enum Entity {
   LOGGINGS = 'Loggings',
   LOGGING = 'Logging',
   PRICING = 'Pricing',
+  PRICING_DEFINITION = 'PricingDefinition',
+  PRICING_DEFINITIONS = 'PricingDefinitions',
```
Comment (LucasBrazi06) [inline src/types/Authorization.ts:88]: Maybe we should have used only singular entities to simplify :-)
But that's another story!

File: src/types/requests/HttpPricingRequest.ts [hunk 17]
```diff
@@ -0,0 +1,15 @@
...
+
+export interface HttpPricingDefinitionRequest extends HttpByIDRequest {
+  ID: string;
+  WithEntityInformation?: boolean;
+}
+
+export interface HttpPricingDefinitionsRequest extends HttpDatabaseRequest {
+  ID?: string;
+  Search?: string;
+  entityID?: string;
```
Comment (LucasBrazi06) [inline src/types/requests/HttpPricingRequest.ts:12]: PC (Pascal Case)

File: src/types/requests/HttpPricingRequest.ts [hunk 18]
```diff
@@ -0,0 +1,15 @@
...
+  entityType?: string;
```
Comment (LucasBrazi06) [inline src/types/requests/HttpPricingRequest.ts:13]: PC

File: src/utils/LoggingHelper.ts [hunk 19]
```diff
@@ -0,0 +1,15 @@
+import { Log } from '../types/Log';
+import Transaction from '../types/Transaction';
+
+export default class LoggingHelper {
+
+  public static getSessionProperties(transaction: Transaction): Partial<Log> {
```
Comment (LucasBrazi06) [inline src/utils/LoggingHelper.ts:6]: Good idea, we can have the same approach for the Charging Station :-).
Another story!

**Review by LucasBrazi06 — APPROVED** [review]
GUT !!!!
I'll merge it, check my comments.
Few changes needed.

File: src/assets/i18n/cz.json [hunk 20]
```diff
@@ -2,7 +2,18 @@
    "billing":{
       "generatedUser":"Generovat uživatele pro '{{email}}'",
       "chargingAtSiteArea":"Nabíjecí relace: {{sessionID}} - počáteční datum: {{startDate}} v {startTime} - spotřeba energie: {{totalConsumption}} kWh v {{siteAreaName}}",
-      "chargingAtChargeBox":"Nabíjecí relace: {{sessionID}} - počáteční datum: {{startDate}} v {{startTime}} - spotřeba energie: {{totalConsumption}} kWh nabíjecí stanice {{chargeBoxID}}"
+      "chargingAtChargeBox":"Nabíjecí relace: {{sessionID}} - počáteční datum: {{startDate}} v {{startTime}} - spotřeba energie: {{totalConsumption}} kWh nabíjecí stanice {{chargeBoxID}}",
```
[the remark this answers is not in the document]
Comment (ClaudeROSSI) [reply src/assets/i18n/cz.json:5]: yes. This text is the one shown in the invoice.

File: src/types/requests/HttpPricingRequest.ts (hunk 17 above)
Comment (ClaudeROSSI) [reply src/types/requests/HttpPricingRequest.ts:12]: @cabch : Let's sync with front-end changes
entityID ==> EntityID
entityType ==> EntityType

**Pull request merged**